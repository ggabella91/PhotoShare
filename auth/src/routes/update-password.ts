import express, { Request, Response } from 'express';
import { body } from 'express-validator';
import jwt from 'jsonwebtoken';

import { Password } from '../services/password';
import { User } from '../models/user';
import { validateRequest, BadRequestError } from '@ggabella-photo-share/common';

const router = express.Router();

router.patch(
  '/api/users/updatePassword',
  [
    body('passwordCurrent')
      .trim()
      .notEmpty()
      .withMessage('You must supply your current password'),
    body('password')
      .trim()
      .isLength({ min: 4, max: 20 })
      .withMessage('Password must be between 4 and 20 characters'),
    body('passwordConfirm')
      .trim()
      .custom((value, { req }) => {
        if (value !== req.body.password) {
          throw new Error('Password confirmation does not match password');
        }

        return true;
      }),
  ],
  validateRequest,
  async (req: Request, res: Response) => {
    const { password, email } = req.body;

    const existingUser = await User.findOne({ email });

    if (!existingUser) {
      throw new BadRequestError('Invalid credentials');
    }

    const passwordsMatch = await Password.compare(
      existingUser.password,
      password
    );

    if (!passwordsMatch) {
      throw new BadRequestError('Invalid credentials');
    }

    existingUser.password = req.body.password;
    await existingUser.save();

    // Generate JWT
    const userJwt = jwt.sign(
      {
        id: existingUser.id,
        email: existingUser.email,
      },
      process.env.JWT_KEY!
    );

    // Store it on the session object
    req.session = {
      jwt: userJwt,
    };

    res.status(200).send(existingUser);
  }
);

export { router as updatePasswordRouter };
