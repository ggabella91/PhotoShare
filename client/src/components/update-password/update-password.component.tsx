import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import {
  changePasswordStart,
  clearPasswordStatuses,
} from '../../redux/user/user.actions';

import { AppState } from '../../redux/root-reducer';

import { FormInput } from '../../components/form-input/form-input.component';
import Button from '../../components/button/button.component';

import Alert from 'react-bootstrap/Alert';

export const UpdatePassword: React.FC = () => {
  const [userPassword, setUserPassword] = useState({
    passwordCurrent: '',
    password: '',
    passwordConfirm: '',
  });

  const [, setShowPassAlert] = useState(true);
  const [statusPass, setStatusPass] = useState({
    success: false,
    error: false,
  });

  const userState = useSelector((state: AppState) => state.user);

  const { changePasswordConfirm, changePasswordError } = userState;
  const dispatch = useDispatch();

  const { passwordCurrent, password, passwordConfirm } = userPassword;

  const handlePasswordChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value, name } = event.target;

    setUserPassword({ ...userPassword, [name]: value });
  };

  const handleSubmitPassword = async (
    event: React.FormEvent<HTMLFormElement>
  ) => {
    event.preventDefault();

    dispatch(
      changePasswordStart({ passwordCurrent, password, passwordConfirm })
    );
  };

  useEffect(() => {
    if (changePasswordError) {
      setStatusPass({ ...statusPass, error: true });
    } else if (changePasswordConfirm) {
      setStatusPass({ ...statusPass, success: true });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [changePasswordError, changePasswordConfirm]);

  const handleRenderAlert = (type: string, message: string) => {
    setTimeout(() => {
      setUserPassword({
        passwordCurrent: '',
        password: '',
        passwordConfirm: '',
      });
      setStatusPass({ success: false, error: false });
      dispatch(clearPasswordStatuses());
    }, 5000);
    return (
      <Alert variant={type} onClose={handleHidePassAlert} dismissible>
        {message}
      </Alert>
    );
  };

  const handleHidePassAlert = () => setShowPassAlert(false);

  return (
    <div className='settings'>
      <form className='change-info' onSubmit={handleSubmitPassword}>
        <span>Change your password</span>
        <FormInput
          type='password'
          name='passwordCurrent'
          value={passwordCurrent}
          onChange={handlePasswordChange}
          label='current password'
        />
        <FormInput
          type='password'
          name='password'
          value={password}
          onChange={handlePasswordChange}
          label='new password'
        />
        <FormInput
          type='password'
          name='passwordConfirm'
          value={passwordConfirm}
          onChange={handlePasswordChange}
          label='confirm new password'
        />
        <div className='button'>
          <Button
            className='submit-button settings-button password-button'
            onClick={handleSubmitPassword}
          >
            <span className='update-info password'>Change Password</span>
          </Button>
        </div>
      </form>
      <div className='settings-alert'>
        {statusPass.error
          ? handleRenderAlert('danger', 'Error changing password.')
          : null}
        {statusPass.success
          ? handleRenderAlert('success', 'Password changed successfully!')
          : null}
      </div>
    </div>
  );
};

export default UpdatePassword;
