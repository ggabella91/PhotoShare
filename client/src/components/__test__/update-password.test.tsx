import { shallow } from 'enzyme';
import React from 'react';
import { UpdatePassword } from '../update-password/update-password.component';

import {
  changePasswordStart,
  clearPasswordStatuses,
} from '../../redux/user/user.actions';

it('renders an update-password component', () => {
  const updatePasswordWrapper = shallow(
    <UpdatePassword
      changePasswordStart={(passwordPayload) =>
        changePasswordStart(passwordPayload)
      }
      changePassError={null}
      changePassConfirm={null}
      clearPasswordStatuses={() => clearPasswordStatuses()}
    />
  );

  expect(updatePasswordWrapper).toMatchSnapshot();
});
