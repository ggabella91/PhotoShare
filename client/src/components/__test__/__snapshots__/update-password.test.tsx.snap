// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an update-password component 1`] = `
<div
  className="settings"
>
  <form
    className="change-info"
    onSubmit={[Function]}
  >
    <span>
      Change your password
    </span>
    <FormInput
      label="current password"
      name="passwordCurrent"
      onChange={[Function]}
      type="password"
      value=""
    />
    <FormInput
      label="new password"
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
    <FormInput
      label="confirm new password"
      name="passwordConfirm"
      onChange={[Function]}
      type="password"
      value=""
    />
    <div
      className="button"
    >
      <Button
        className="submit-button settings-button password-button"
        onClick={[Function]}
      >
        <span
          className="update-info password"
        >
          Change Password
        </span>
      </Button>
    </div>
  </form>
  <div
    className="settings-alert"
  />
</div>
`;
